File: utils/Queue.js
module.exports = class Queue {
  constructor () {
    this.stack1 = [];
    this.stack2 = [];
    this.length = 0;
  }

  push(item) {
    this.stack1.push(item);
    this.length += 1;
  }

  pop() {
    this.length -= 1;
    if (this.stack2.length) {
      return this.stack2.pop();
    } else {
      while (this.stack1.length) {
        this.stack2.push(this.stack1.pop())
      }
      return this.stack2.pop()
    }
  }
}

File: ThunderMarket/Models/Map/Queue.swift
//
//  Queue.swift
//  ThunderMarket
//
//  Created by 윤상진 on 2021/08/22.
//

import Foundation

struct Queue<T> {
    var stack1 = [T]()
    var stack2 = [T]()
    var count: Int {
        return stack1.count + stack2.count
    }
    
    var top: T? {
        if let top = stack2.last {
            return top
        }
        if let top = stack1.first {
            return top
        }
        return nil
    }
    
    mutating func append(_ item: T) {
        stack1.append(item)
    }
    
    @discardableResult
    mutating func pop() -> T? {
        if let lastItem = stack2.popLast() {
            return lastItem
        }
        while let lastItem = stack1.popLast() {
            stack2.append(lastItem)
        }
        if let lastItem = stack2.popLast() {
            return lastItem
        }
        return nil
    }
}
